src/db.c

```diff
@@ -980,13 +981,24 @@ void randomkeyCommand(client *c) {
 void keysCommand(client *c) {
     dictEntry *de;
     sds pattern = c->argv[1]->ptr;
-    int plen = sdslen(pattern), allkeys;
+    int plen = sdslen(pattern), allkeys, pslot = -1;
     long numkeys = 0;
     void *replylen = addReplyDeferredLen(c);
-    dbIterator *dbit = dbIteratorInit(c->db, DB_MAIN);
     allkeys = (pattern[0] == '*' && plen == 1);
+    dbIterator *dbit;
+    if (server.cluster_enabled) {
+        if (!allkeys) {
+            pslot = patternHashSlot(pattern, plen);
+        }
+        dbit = dbIteratorInitFromSlot(c->db, DB_MAIN, pslot > -1 ? pslot - 1 : -1);
+    } else {
+        dbit = dbIteratorInit(c->db, DB_MAIN);
+    }
     robj keyobj;
     while ((de = dbIteratorNext(dbit)) != NULL) {
+        if (pslot > -1 && dbIteratorGetCurrentSlot(dbit) > pslot) {
```

Comment:
Just personal (minor) preference

```suggestion
        if (pslot >= 0 && dbIteratorGetCurrentSlot(dbit) > pslot) {
```